software/bsg_manycore_lib/bsg_mcs_mutex.S

```diff
@@ -0,0 +1,80 @@
+//
+// (2) The lock approximates a FIFO-ish structure, which improves fairness. A simple spinlock on the manycore
+// will favor threads topologically closer to the memory bank in which the lock resides and can lead to
+// starvation of the other cores.
+//
+// This lock is by no means perfect. For locks with low contention, a simple spinlock may result in better performance.
+
+        .text
+        .globl bsg_mcs_mutex_acquire
+bsg_mcs_mutex_acquire:
+        // a0 = mtx         : bsg_mcs_mutex_t*
+        // a1 = lcl         : bsg_mcs_mutex_node_t*
+        // a2 = lcl_as_glbl : bsg_mcs_mutex_node_t*
+        sw      x0, 4(a1)                      // lcl->next = 0
+        sw      x0, 0(a1)                      // lcl->unlocked = 0
+        amoswap.w.aq t0, a2, 0(a0)             // predecessor = swap (&mtx, lcl_as_glbl)
+        beqz    t0, bsg_mcs_mutex_acquire_ret  // return if predecessor = 0
+        sw      a2, 4(t0)                      // predecessor->next = lcl_as_glbl
+        lr.w    t0, (a1)                       // unlocked = lcl->unlocked
+        bnez    t0, bsg_mcs_mutex_acquire_ret  // return if unlocked
+bsg_mcs_mutex_acquire_loop:
+        lr.w.aq t0, (a1)                       // unlocked = lcl->unlocked
+        beqz    t0, bsg_mcs_mutex_acquire_loop // while !unlocked
+bsg_mcs_mutex_acquire_ret:
+        ret
+
+        .globl bsg_mcs_mutex_release
+bsg_mcs_mutex_release:
+        // a0 = mtx         : bsg_mcs_mutex_t*
+        // a1 = lcl         : bsg_mcs_mutex_node_t*
+        // a2 = lcl_as_glbl : bsg_mcs_mutex_node_t*
+        lw      t0, 4(a1)                                   // next = lcl->next
+        li      t1, 1                                       // t1 = 1
+        beqz    t0, bsg_mcs_mutex_release_no_successor      // branch if no successor
+        // this is the case where there is a successor
+        // we need only unlock the successor and return
+        fence                                               // fence to implement release semantics
+        sw      t1, 0(t0)                                   // successor->unlocked = 1
+        ret
+bsg_mcs_mutex_release_no_successor:
+        // this is the case where there is no known successor
+        // attempt to swap out the tail pointer with 0
+        amoswap.w.rl t2, x0, 0(a0)                          // victim_tail = swap(&mtx, 0)
+        bne     t2, a2, bsg_mcs_mutex_release_exists_victim // victim_tail == lcl_as_glbl?
+        ret                                                 // there really is no successor -- return
+bsg_mcs_mutex_release_exists_victim:
+        // someone added themselves to the queue and we have removed them
+        // we need to put them back
+        amoswap.w t2, t2, 0(a0)                             // usurper = swap(&mtx, victim_tail)
+bsg_mcs_mutex_release_wait_on_successor:
+        lw      t0, 4(a1)                                   // next = lcl->next
```

Comment:
should be using LBR?

Reply:
I guess I made a judgement call here. My reasoning was that this situation means a successor is in the process of acquiring the lock and an update to the next pointer is imminent (10s of cycles). Using LBR requires extra instructions and I thought it was likely not worth the additional instruction footprint given that this a hopefully uncommon corner case.

Reply:
maybe you can document this in the code? for each of these synchronizations, what is the expectation?

Reply:
done